### PR TITLE
ledger refactoring: resourceData tests

### PR DESCRIPTION
**Pull request by algorandskiy**
Added a test matrix for all possible resourceData and resourceDelta states

### CODE REVIEW

File: ledger/accountdb.go [hunk 1]
```diff
@@ -1402,14 +1402,18 @@ func (rd *resourcesData) ClearAssetHolding() {
 		rd.ResourceFlags &= ^resourceFlagsEmptyAsset
 	}
 	rd.ResourceFlags |= resourceFlagsNotHolding
+	hadParams := (rd.ResourceFlags & resourceFlagsOwnership) == resourceFlagsOwnership
```
Comment (tsachiherman) [inline ledger/accountdb.go:1405]: suggested cleanup:
```golang
func (rd *resourcesData) ClearAssetHolding() {
	rd.Amount = 0
	rd.Frozen = false

	rd.ResourceFlags |= resourceFlagsNotHolding
	hadParams := (rd.ResourceFlags & resourceFlagsOwnership) == resourceFlagsOwnership
	if hadParams && rd.IsEmptyAssetFields() {
		rd.ResourceFlags |= resourceFlagsEmptyAsset
	} else {
		rd.ResourceFlags &= ^resourceFlagsEmptyAsset
	}
}
```

File: ledger/accountdb.go [hunk 2]
```diff
@@ -1432,14 +1436,18 @@ func (rd *resourcesData) ClearAppLocalState() {
 		rd.ResourceFlags &= ^resourceFlagsEmptyApp
 	}
 	rd.ResourceFlags |= resourceFlagsNotHolding
+	hadParams := (rd.ResourceFlags & resourceFlagsOwnership) == resourceFlagsOwnership
```
Comment (tsachiherman) [inline ledger/accountdb.go:1439]: suggested cleanup:
```golang
func (rd *resourcesData) ClearAppLocalState() {
	rd.SchemaNumUint = 0
	rd.SchemaNumByteSlice = 0
	rd.KeyValue = nil

	rd.ResourceFlags |= resourceFlagsNotHolding
	hadParams := (rd.ResourceFlags & resourceFlagsOwnership) == resourceFlagsOwnership
	if hadParams && rd.IsEmptyAppFields() {
		rd.ResourceFlags |= resourceFlagsEmptyApp
	} else {
		rd.ResourceFlags &= ^resourceFlagsEmptyApp
	}
}
```

File: ledger/accountdb.go [hunk 3]
```diff
@@ -339,7 +339,7 @@ func prepareNormalizedBalancesV6(bals []encodedBalanceRecordV6, proto config.Con
 				return nil, err
 			}
 			ctype := basics.AssetCreatable
-			if resData.IsEmptyApp() {
+			if resData.IsEmptyAppFields() {
```
Comment (tsachiherman) [inline ledger/accountdb.go:342]: Could you please change this one to `resData.IsEmptyAppFields()` -> `resData.IsApp()` ?
This was the culprit for mismatching digests..

Comment (algorandskiy) [reply ledger/accountdb.go]: done

**Review by tsachiherman — APPROVED** [review]
looks great!

**Pull request merged**